packages/ListBoxBrowser/src/helpers.js

```diff
@@ -0,0 +1,259 @@
+import React from "react";
+
+/** helpers */
+const hasOptions = options => typeof options !== "undefined" && Array.isArray(options);
+
+export function isRoot(key) {
+  if (key === null) return false;
+
+  return key.includes("root");
+}
+
+function getOptions(indexes, list) {
+  if (Array.isArray(indexes)) {
+    return indexes.map(index => {
+      return list[index].value;
+    });
+  }
+
+  return [list[indexes].value];
+}
+
+function isSelectable({ hasOptions, isParentSelectable }) {
+  if (isParentSelectable !== null) {
+    return !isParentSelectable;
+  }
+
+  return hasOptions;
+}
+
+/** DATA STRUCTURE */
+
+export function getData({
+  data,
+  path = "root",
+  refDefaultSelectedOptions,
+  defaultSelectedOptions,
+  defaultSelectedView,
+  refDefaultSelectedView,
+}) {
+  return data.map(({ label = null, options, ...moreAttributes }, index) => {
+    if (!label) {
+      throw new Error("A Label attribute is required for each option object, can't process data.");
+    }
+
+    let newPath = isRoot(path) ? `${index}` : `${path}/${index}`;
+    const recursiveParameters = {
+      data: options,
+      defaultSelectedOptions,
+      defaultSelectedView,
+      path: newPath,
+      refDefaultSelectedOptions,
+      refDefaultSelectedView,
+    };
+    const _hasOptions = hasOptions(options);
+
+    const option = {
+      parent: path,
+      hasOptions: _hasOptions,
+      $$key: newPath,
+      attributes: {
+        label,
+        options: _hasOptions ? getData(recursiveParameters) : null,
+        ...moreAttributes,
+      },
+    };
+
+    if (defaultSelectedOptions(option.attributes)) {
+      /* eslint-disable no-param-reassign */
+      if (isRoot(option.parent)) {
+        newPath = "root";
+      }
+
+      if (option.parent in refDefaultSelectedOptions.current) {
+        refDefaultSelectedOptions.current = {
+          ...refDefaultSelectedOptions.current,
+          [option.parent]: [...refDefaultSelectedOptions.current[option.parent], option],
+        };
+      } else {
+        refDefaultSelectedOptions.current = { ...refDefaultSelectedOptions.current, [option.parent]: [option] };
+      }
+      /* eslint-enable no-param-reassing */
+    }
+
+    // checking if its initial view
+    if (_hasOptions && defaultSelectedView && defaultSelectedView(option.attributes)) {
+      refDefaultSelectedView.current = newPath;
```

Comment:
what happens if there are more than one option returned, it will just set the default view to the latest iterated option?, perhaps if this is set more than once based on the data there should be an error... 🤷‍♂

Reply:
will get the latest options selected, not a big deal this is handled by the consumer so it's up to them to return the correct option.